authorization-common/src/main/java/com/sequenceiq/authorization/service/EnvironmentBasedAuthorizationProvider.java

```diff
@@ -34,61 +35,61 @@
-                    "thus authorization is failing automatically.", action, Joiner.on(",").join(
-                            resourceCrns.stream().map(crn -> Crn.safeFromString(crn).getResourceType().name()).collect(Collectors.toSet())));
-            throw new AccessDeniedException(String.format("Action %s is not supported over resources %s, thus access is denied",
-                    action.getRight(), Joiner.on(",").join(resourceCrns)));
-        }
-        Map<String, Optional<String>> withEnvironmentCrn = resourceBasedCrnProvider.getEnvironmentCrnsByResourceCrns(resourceCrns);
-        if (MapUtils.isEmpty(withEnvironmentCrn)) {
-            return HasRightOnAll.from(action, resourceCrns);
+        if (action.getAuthorizationResourceType().isHierarchicalAuthorizationNeeded()) {
+            AuthorizationEnvironmentCrnListProvider environmentCrnListProvider = environmentCrnListProviderMap.get(action.getAuthorizationResourceType());
+            if (environmentCrnListProvider == null) {
+                LOGGER.error("There is no resource based crn provider implemented for action {} against resource types {}, " +
+                        "thus authorization is failing automatically.", action, Joiner.on(",").join(
+                        resourceCrns.stream().map(crn -> Crn.safeFromString(crn).getResourceType().name()).collect(Collectors.toSet())));
+                throw new AccessDeniedException(String.format("Action %s is not supported over resources %s, thus access is denied",
+                        action.getRight(), Joiner.on(",").join(resourceCrns)));
+            }
+            Map<String, String> withEnvironmentCrn = environmentCrnListProvider.getEnvironmentCrnsByResourceCrns(resourceCrns);
+            if (MapUtils.isEmpty(withEnvironmentCrn)) {
+                return HasRightOnAll.from(action, resourceCrns);
+            } else {
+                Map<String, Set<String>> byEnvironments = new LinkedHashMap<>();
+                withEnvironmentCrn.forEach((resourceCrn, envCrn) ->
+                        byEnvironments.computeIfAbsent(envCrn, s -> new LinkedHashSet<>()).add(resourceCrn));
+                List<AuthorizationRule> authorizations = new ArrayList<>();
+                authorizations.addAll(byEnvironments.entrySet()
+                        .stream()
+                        .map(e -> {
+                            if (e.getValue().size() == 1) {
+                                return new HasRightOnAny(action, List.of(e.getKey(), e.getValue().iterator().next()));
+                            } else {
+                                return new AnyMatch(List.of(
+                                        new HasRight(action, e.getKey()),
+                                        new HasRightOnAll(action, e.getValue())
+                                ));
+                            }
+                        }).collect(Collectors.toList()));
+                return AllMatch.fromList(authorizations);
```

Comment:
Can you please split these into smaller methods? It's a bit hard to grasp what is happening.

Reply:
sure